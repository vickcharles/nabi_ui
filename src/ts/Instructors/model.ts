export interface ExperienceType {
  id?: number;
  employer: string;
  jobTitle: string;
  fromMonth: string;
  fromYear: string;
  toMonth: string;
  toYear: string;
  location: string;
}

export enum DegreeType {
  associate = 'Associate Degree',
  bachelors = 'Bachelor\'s Degree',
  graduate = 'Graduate Degreee',
  professional = 'Professional Degree',
  certification = 'Certification',
  other = 'Other'
}

export interface EducationType {
  id?: number;
  schoolName: string;
  graduationYear: string;
  degreeType: DegreeType;
  fieldOfStudy: string;
}

export enum PlaceForLessonsOptions {
  home = 'Student\'s home',
  studio = 'Instructor\'s studio',
  online = 'Online',
}

export interface PlaceForLessonsType {
  home: boolean;
  studio: boolean;
  online: boolean;
}

export enum LessonDuration {
  thirtyMins = '30 mins',
  fortyFiveMins = '45 mins',
  sixtyMins = '60 mins',
  ninetyMins = '90 mins'
}

export interface RatesType {
  thirtyMinsRate: number;
  fortyFiveMinsRate: number;
  sixtyMinsRate: number;
  ninetyMinsRate: number;
}

export enum SkillLevel {
  beginner = 'Beginner',
  intermediate = 'Intermediate',
  advanced = 'Advanced'
}

export interface InstrumentsType {
  instrument: string;
  skillLevel: SkillLevel;
}

export interface InstructorType {
  userId: string;
  bio?: string;
  instruments?: InstrumentsType[];
  rates?: RatesType;
  placeForLessons?: PlaceForLessonsType;
  studioAddress?: string;
  education?: EducationType[];
  experience?: ExperienceType[];
}

export type IState = InstructorType[];
