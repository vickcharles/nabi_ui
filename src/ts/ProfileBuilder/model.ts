export interface InstructorState {
  userId: string;
  bio?: string;
}

export type IState = InstructorState[];
  