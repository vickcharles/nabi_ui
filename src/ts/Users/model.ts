export enum HearAboutUs {
  Google = 'google',
  Facebook = 'facebook',
  Craigslist = 'craigslist',
  PrintAds = 'printAds',
  WordOfMouth = 'wordOfMouth',
  CurrentNabiUser = 'currentNabiUser',
  Other= 'other'
}

export interface HearAboutUsInfo {
  value: HearAboutUs;
  label: string;
}

export const hearAboutUsInfo = {
  Google: <HearAboutUsInfo> {
    value: HearAboutUs.Google,
    label: 'Google'
  },
  Facebook: <HearAboutUsInfo> {
    value: HearAboutUs.Facebook,
    label: 'Facebook'
  },
  Craigslist: <HearAboutUsInfo> {
    value: HearAboutUs.Craigslist,
    label: 'Craigslist'
  },
  PrintAds: <HearAboutUsInfo> {
    value: HearAboutUs.PrintAds,
    label: 'Print Ads'
  },
  WordOfMouth: <HearAboutUsInfo> {
    value: HearAboutUs.WordOfMouth,
    label: 'Word Of Mouth'
  },
  CurrentNabiUser: <HearAboutUsInfo> {
    value: HearAboutUs.CurrentNabiUser,
    label: 'Current Nabi User'
  },
  Other: <HearAboutUsInfo> {
    value: HearAboutUs.Other,
    label: 'Other'
  }
};

export enum Role {
  student = 'student',
  instructor = 'instructor',
  parent = 'parent'
}

export interface UserState {
  id: string;
  firstName: string;
  lastName: string;
  email: string;
  password: string;
  zipCode: string;
  role: Role;
  hearAboutUs: string;
  avatar?: string;
  displayName: string;
}

export type IState = UserState[];
