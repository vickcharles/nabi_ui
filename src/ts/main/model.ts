/**
 * State for firing redirect
 * @interface RedirectState
 */
export interface RedirectState {
  fireRedirect: boolean;
}
