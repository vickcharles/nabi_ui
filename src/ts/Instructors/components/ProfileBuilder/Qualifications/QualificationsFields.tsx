import * as React from 'react';
import Typography from 'material-ui/Typography';
import { FormGroup, FormControlLabel } from 'material-ui/Form';
import Checkbox from 'material-ui/Checkbox';
import Grid from 'material-ui/Grid';

import { qualificationsOptions, QualificationsType } from '../../../model';

interface QualificationsFieldsProps extends  QualificationsType {
  handleChangeQualifications: (event: React.FormEvent<{}>) => void;
}

const QualificationsFields: React.StatelessComponent<QualificationsFieldsProps> = props => {
  const qualificationFields: any = [];

  for (const [key, value] of Object.entries(qualificationsOptions)) {
    qualificationFields.push(
      <FormControlLabel
        key={key}
        control={
          <Checkbox
            name={value.name}
            onChange={props.handleChangeQualifications}
          />
        }
        label={value.label}
      />
    );
  }

  const qualificationFieldsFirstHalf = qualificationFields.slice(0, 5);
  const qualificationFieldsSecondHalf = qualificationFields.slice(5);
  return (
    <div>
      <Typography className="nabi-margin-top-small nabi-margin-bottom-small" variant="body2">
        Additional Qualifications
      </Typography>
      <FormGroup className="nabi-margin-left-small">
        <Grid container={true} spacing={24}>
          <Grid className="qualifications nabi-padding-top-zero nabi-padding-bottom-zero" item={true} xs={12} md={6}>
            {qualificationFieldsFirstHalf}
          </Grid>
          <Grid className="qualifications nabi-padding-top-zero" item={true} xs={12} md={6}>
          {qualificationFieldsSecondHalf}
          </Grid>
        </Grid>
      </FormGroup>
    </div>
  );
};

export default (QualificationsFields);
