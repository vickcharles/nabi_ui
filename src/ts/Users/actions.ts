import { Action, Dispatch } from 'redux';
import { UserActions, ZipCodeActions } from './constants/ActionTypes';
import { UserState } from './model';
import axios  from 'axios';

interface CreateUser extends Action {
  user: UserState;
}

interface ChangeAvatar extends Action {
  id: string;
  avatar: string;
}

interface UpdateUser extends Action {
  user: UserState;
}

interface ZipCodeFetch extends Action {
  zipcodedata: any;
}

// Create user
export function createUser(user: UserState): CreateUser {
  return {
    user,
    type: UserActions.CREATE_USER
  };
}

/**
 * Action to change avatar
 * 
 * @param {string} id - The id of the user
 * @param {string} avatar - The avatar of the user
 */
export function changeAvatar(id: string, avatar: string): ChangeAvatar {
  return {
    id: id,
    avatar: avatar ,
    type: UserActions.CHANGE_AVATAR
  };
}

/**
 * Action to update user
 * 
 * @param {string} user - The new user
 */
export function updateUser(user: UserState): UpdateUser {
  return {
    user: user,
    type: UserActions.UPDATE_USER
  };
}

function fetch_zip_start(): Action {
 return {
   type: ZipCodeActions.FETCH_ZIPADDRESS_START
 };
}

function fetch_zip_done(city: string, state: string, country: string, id: string): ZipCodeFetch {
 return {
   zipcodedata: { city: city, state: state, country: country, id: id},
   type: ZipCodeActions.FETCH_ZIPADDRES_DONE
 };
}

function fetch_zip_error(): Action {
 return {
   type: ZipCodeActions.FETCH_ZIPADDRESS_ERROR
 };
}

/**
 * Special Action for Fetching with axios, Gets user zipcode and dispatches
 * 3 actions, Starting, Done Fetching, Some Error. This way is best practice
 * to load data from API so UI doesn't depend on internal calls. 
 * 
 * @export
 * @param {UserState} user 
 * @returns 
 */
export function fetchZipCodeAddress( user: UserState ) {
 return async (dispatch: Dispatch<UserState>): Promise<void> => {
   dispatch(fetch_zip_start());
   let url = `https://maps.googleapis.com/maps/api/geocode/json?`;
   url += `address=${user.zipCode}&key=AIzaSyDFRiJ2j-4brJg2_MLmxZ-TXK4ObZee69c`;
   return axios.get(url)
   .then(
     (res) => {
        let AddressForm = res.data.results[0] || [];
        if ( res.data.results.length ) {
          function getValues(values: string[]) {
            return AddressForm.address_components.filter( (obj: any) => {
              return obj.types.some((o: any) => values.includes(o));
            });
          }
          user.city = getValues(['neighborhood', 'locality', 'sublocality'])[0].short_name;
          user.state = getValues(['administrative_area_level_1'])[0].short_name;
          user.country = getValues(['country'])[0].short_name;
          dispatch( fetch_zip_done( user.city || ''  , user.state || '' , user.id || '', user.country || '' ));
       } else {
          dispatch( fetch_zip_done( '', '', '', user.id ) );
       }
     }
   )
   .catch(
     (err) => {
        dispatch(fetch_zip_error());
     } 
   );
 };
}
