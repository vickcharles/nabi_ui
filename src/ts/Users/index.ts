export { default as Registration } from './components/Registration';
export * from './actions';
import * as User from './model';
export { User };
import reducer from './reducer';
export default reducer;
