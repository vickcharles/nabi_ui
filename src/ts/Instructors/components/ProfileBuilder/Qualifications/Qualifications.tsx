import * as React from 'react';
import { Action, Dispatch } from 'redux';
import { ThunkAction } from 'redux-thunk';
import { connect } from 'react-redux';

import { UserState } from '../../../../Users/model';
import { updateUser } from '../../../../Users';
import { InstructorType, QualificationsType } from '../../../model';
import { updateInstructor } from '../../../';
import QualificationsFields from './QualificationsFields';

/**
 * Qualifications dispatch props
 * @interface QualificationsDispatchProps
 */
interface QualificationsDispatchProps {
  updateInstructor: (instructor: InstructorType) => void;
  updateUser: (user: UserState) => void;
}

/**
 * Qualifications's own props
 * @interface QualificationsOwnProps
 */
interface QualificationsOwnProps {
  userId: string;
}

/**
 * State for Qualifications
 * @interface QualificationsState
 * @extends QualificationsType
 */
interface QualificationsState extends QualificationsType {}

/**
 * Props for Qualifications
 * @interface QualificationsProps
 * @extends QualificationsDispatchProps & QualificationsOwnProps
 */
interface QualificationsProps extends
  QualificationsDispatchProps,
  QualificationsOwnProps { }

/**
 * Qualifications: Fourth step of the Profile Builder
 * @class Qualifications
 * @extends React.Component<QualificationsProps, QualificationsState>
 */
export class Qualifications extends React.Component<QualificationsProps, QualificationsState> {
  constructor(props: QualificationsProps) {
    super(props);

    this.state = {
      add: false,
      aspergerSyndrome: false,
      autism: false,
      depression: false,
      diabetes: false,
      downSyndrome: false,
      epilepsy: false,
      musicTherapy: false,
      ocd: false,
      visualImpairment: false,
    };

    this.handleChangeQualifications = this.handleChangeQualifications.bind(this);
  }

  public updateInstructorCall(instructor: InstructorType): void {
    this.props.updateInstructor(instructor);
  }

  public handleChangeQualifications(event: any): void {
    const target = event.target;
    const name = target.name;

    if (target.checked) {
      this.setState({ [name]: true }, () => {
        const qualifications = {
          add: this.state.add,
          aspergerSyndrome: this.state.aspergerSyndrome,
          autism: this.state.autism,
          depression: this.state.depression,
          diabetes: this.state.diabetes,
          downSyndrome: this.state.downSyndrome,
          epilepsy: this.state.epilepsy,
          musicTherapy: this.state.musicTherapy,
          ocd: this.state.ocd,
          visualImpairment: this.state.visualImpairment
        };

        const instructor: InstructorType =  {
          userId: this.props.userId,
          qualifications: qualifications
        };
        this.updateInstructorCall(instructor);
      });
    } else {
      this.setState({ [name]: false }, () => {
        const qualifications = {
          add: this.state.add,
          aspergerSyndrome: this.state.aspergerSyndrome,
          autism: this.state.autism,
          depression: this.state.depression,
          diabetes: this.state.diabetes,
          downSyndrome: this.state.downSyndrome,
          epilepsy: this.state.epilepsy,
          musicTherapy: this.state.musicTherapy,
          ocd: this.state.ocd,
          visualImpairment: this.state.visualImpairment
        };

        const instructor: InstructorType =  {
          userId: this.props.userId,
          qualifications: qualifications,
        };
        this.updateInstructorCall(instructor);
      });
    }
  }

  public render(): JSX.Element {
    return (
      <div>
        <QualificationsFields
          handleChangeQualifications={this.handleChangeQualifications}
          add={this.state.add}
          aspergerSyndrome={this.state.aspergerSyndrome}
          autism={this.state.autism}
          depression={this.state.depression}
          diabetes={this.state.diabetes}
          downSyndrome={this.state.downSyndrome}
          epilepsy={this.state.epilepsy}
          musicTherapy={this.state.musicTherapy}
          ocd={this.state.ocd}
          visualImpairment={this.state.visualImpairment}
        />
      </div>
    );
  }
}

function mapDispatchToProps(
  dispatch: Dispatch<Action | ThunkAction<{}, {}, {}>>,
  _ownProps: QualificationsOwnProps
): QualificationsDispatchProps {
  return {
    updateInstructor: (instructor: InstructorType) => dispatch(updateInstructor(instructor)),
    updateUser: (user: UserState) => dispatch(updateUser(user))
  };
}

export default connect(null, mapDispatchToProps)(Qualifications);
