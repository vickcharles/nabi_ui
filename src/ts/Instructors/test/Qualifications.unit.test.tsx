import * as React from 'react';
import { Qualifications } from '../components/ProfileBuilder/Qualifications/Qualifications';
import { shallow } from 'enzyme';
import { InstructorType } from '../model';
import { UserState } from '../../Users/model';

describe('Qualifications', () => {
  let wrapper: any;
  const updateInstructor: (instructor: InstructorType) => {} = jest.fn();
  const updateUser: (user: UserState) => {} = jest.fn();
  const updateInstructorCall = jest.spyOn(Qualifications.prototype, 'updateInstructorCall');

  describe('Shallow', () => {
    beforeEach(() => {
      wrapper = shallow(
        <Qualifications
          userId=""
          updateInstructor={updateInstructor}
          updateUser={updateUser}
        />
      );
    });

    it('Matches snapshot', () => {
      expect(wrapper).toMatchSnapshot();
    });
  });

  describe('Method handleChangeQualifications()', () => {
    function test_handleChangeQualifications(theName: string, theValue: boolean): void {
      describe(`When the event contains ${theName} for name and ${theValue} for value`, () => {
        beforeEach(() => {
          const e = {
            target: {
              name: theName,
              checked: theValue
            }
          };

          wrapper.instance().handleChangeQualifications(e);
        });

        it(`Sets the state's ${theName} to ${theValue}`, () => {
          expect(wrapper.state(theName)).toBe(theValue);
        });

        it('Call updateInstructorCall', () => {
          expect(updateInstructorCall).toBeCalled();
        });
      });
    }

    test_handleChangeQualifications('schoolName', true);
    test_handleChangeQualifications('graduationYear', false);
  });
});
