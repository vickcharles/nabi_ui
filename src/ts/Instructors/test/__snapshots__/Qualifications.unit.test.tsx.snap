// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Qualifications Shallow Matches snapshot 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Qualifications
    updateInstructor={[MockFunction]}
    updateUser={[MockFunction]}
    userId=""
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": <QualificationsFields
        add={false}
        aspergerSyndrome={false}
        autism={false}
        depression={false}
        diabetes={false}
        downSyndrome={false}
        epilepsy={false}
        handleChangeQualifications={[Function]}
        musicTherapy={false}
        ocd={false}
        visualImpairment={false}
      />,
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "add": false,
        "aspergerSyndrome": false,
        "autism": false,
        "depression": false,
        "diabetes": false,
        "downSyndrome": false,
        "epilepsy": false,
        "handleChangeQualifications": [Function],
        "musicTherapy": false,
        "ocd": false,
        "visualImpairment": false,
      },
      "ref": null,
      "rendered": null,
      "type": [Function],
    },
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": <QualificationsFields
          add={false}
          aspergerSyndrome={false}
          autism={false}
          depression={false}
          diabetes={false}
          downSyndrome={false}
          epilepsy={false}
          handleChangeQualifications={[Function]}
          musicTherapy={false}
          ocd={false}
          visualImpairment={false}
        />,
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "add": false,
          "aspergerSyndrome": false,
          "autism": false,
          "depression": false,
          "diabetes": false,
          "downSyndrome": false,
          "epilepsy": false,
          "handleChangeQualifications": [Function],
          "musicTherapy": false,
          "ocd": false,
          "visualImpairment": false,
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
